### DISCUSSION TITLE
How to integrate pyecharts?

### DISCUSSION THREAD

**Discussion by SparkBayes** [zauberzeug/nicegui#1701] [Q&A]
### Question

pyecharts：https://github.com/pyecharts/pyecharts-gallery/blob/master/README_EN.md

**Comment by falkoschindler** [comment]
https://pyecharts.org/#/en-us/

I'm not sure. At a first glance, pyecharts allows to create charts like this:

```py
from pyecharts.charts import Bar

bar = Bar()
bar.add_xaxis(["shirts", "cardigans", "chiffons", "trousers", "heels", "socks"])
bar.add_yaxis("Merchant A", [5, 20, 36, 10, 75, 90])
bar.render()
```

> `render` will generate a local HTML file, by default render.html will be generated in the current directory.
> You can also pass in a path parameter, e.g. bar.render("mycharts.html").

You can also use `bar.render_embed()` to get a string of the rendered HTML. But since the result is the whole HTML page including html, head and body tag, I'm not sure how to easily extract individual components for `ui.add_head_html`, `ui.add_body_html` and `ui.html`. Maybe there's a way to avoid the rendering step and extract the options dictionary for `ui.echart` directly?

**Comment by Xiaotian13** [comment]
Since the key is to extract the corresponding options from pyecharts, you might want to try this:
```python
from json import loads
ui.echart(options=loads(bar.dump_options_with_quotes())
```
The variable `bar` is a chart created from pyecharts. In my case it works.

**Reply by falkoschindler** [reply]
Nice! 😃

In combination with my example:
```py
import json
from pyecharts.charts import Bar
from nicegui import ui

bar = Bar()
bar.add_xaxis(["shirts", "cardigans", "chiffons", "trousers", "heels", "socks"])
bar.add_yaxis("Merchant A", [5, 20, 36, 10, 75, 90])
bar.render()

ui.echart(json.loads(bar.dump_options_with_quotes()))

ui.run()
```

**Reply by deepindeep491** [reply]
Thanks, @Xiaotian13 , @falkoschindler . This works fine with Bar chart, but seems to not working with Map chart from pyecharts?
In my case nothing is displayed on the page. Here is the example code:
```py
import json
from pyecharts import options as opts
from pyecharts.charts import Map
from pyecharts.faker import Faker
from nicegui import ui

worldmap = Map()
worldmap.add("V", [list(z) for z in zip(Faker.country, Faker.values())], "world")
worldmap.set_series_opts(label_opts=opts.LabelOpts(is_show=False))
worldmap.set_global_opts(
        title_opts=opts.TitleOpts(title="World Map"),
        visualmap_opts=opts.VisualMapOpts(max_=200),
    )
worldmap.render()

ui.echart(json.loads(worldmap.dump_options_with_quotes()))

ui.run()
```

**Reply by falkoschindler** [reply]
@deepindeep491 There's an error on the JavaScript console:
> TypeError: Cannot read properties of undefined (reading 'regions')

Maybe this is related: https://github.com/apache/echarts/issues/17291

**Reply by deepindeep491** [reply]
@falkoschindler Thanks for the reply! I'm new to python and don't know about echarts js lib. 
What I want to do with nicegui is to display a interactive map from local GeoJson files, is there any way to achieve this?

**Reply by falkoschindler** [reply]
@deepindeep491 You can try using NiceGUI's map element `ui.leaflet` for rendering GeoJSON:

```py
geojson_feature = {
    'type': 'Feature',
    'properties': {'name': 'Coors Field'},
    'geometry': {'type': 'Point', 'coordinates': [-104.99404, 39.75621]}
}
m = ui.leaflet(center=(39.75621, -104.99404)).classes('h-96')
m.generic_layer(name='geoJSON', args=[geojson_feature])
```

This is only a proof of concept. Please refer to [`ui.leaflet`](https://nicegui.io/documentation/leaflet) and the Leaflet and GeoJSON documentation for more details.

**Reply by Hazedd** [reply]
@falkoschindler thx! helpt me out was trying to add the whole feature collection, then it's just adding 1 feature... 

anny change you know a way to get the click on feature event to work ?

**Reply by falkoschindler** [reply]
@Hazedd Sorry, I can't find a way to subscribe to click events on a feature.
